src/planner.c

```diff
@@ -523,16 +393,95 @@ involves_hypertable(PlannerInfo *root, RelOptInfo *rel)
 	}
 }
 
+
 /*
+ * Customize INSERT (ModifyTablePath) paths on hypertables.
+ *
+ * From the ModifyTable description: "Each ModifyTable node contains
+ * a list of one or more subplans, much like an Append node.  There
+ * is one subplan per result relation."
+ *
+ * The subplans produce the tuples for INSERT, while the result relation is the
+ * table we'd like to insert into.
+ *
+ * The way we redirect tuples to chunks is to insert an intermediate "chunk
+ * dispatch" plan node, inbetween the ModifyTable and its subplan that produces
+ * the tuples. When the ModifyTable plan is executed, it tries to read a tuple
+ * from the intermediate chunk dispatch plan instead of the original
+ * subplan. The chunk plan reads the tuple from the original subplan, looks up
+ * the chunk, sets the executor's resultRelation to the chunk table and finally
+ * returns the tuple to the ModifyTable node.
+ *
+ * We also need to wrap the ModifyTable plan node with a HypertableInsert node
+ * to give the ChunkDispatchState node access to the ModifyTableState node in
+ * the execution phase.
+ *
+ * Conceptually, the plan modification looks like this:
+ *
+ * Original plan:
+ *
+ *		  ^
+ *		  |
+ *	[ ModifyTable ] -> resultRelation
+ *		  ^
+ *		  | Tuple
+ *		  |
+ *	  [ subplan ]
+ *
+ *
+ * Modified plan:
+ *
+ *	[ HypertableInsert ]
+ *		  ^
+ *		  |
+ *	[ ModifyTable ] -> resultRelation
+ *		  ^			   ^
+ *		  | Tuple	  / <Set resultRelation to the matching chunk table>
+ *		  |			 /
+ * [ ChunkDispatch ]
+ *		  ^
+ *		  | Tuple
+ *		  |
+ *	  [ subplan ]
+ *
  * PG11 adds a value to the create_upper_paths_hook for FDW support. (See:
  * https://github.com/postgres/postgres/commit/7e0d64c7a57e28fbcf093b6da9310a38367c1d75).
  * Additionally, it calls the hook in a different place, once for each
  * RelOptInfo (see:
  * https://github.com/postgres/postgres/commit/c596fadbfe20ff50a8e5f4bc4b4ff5b7c302ecc0),
  * we do not change our behavior yet, but might choose to in the future.
  */
-static
-void
+static List *
+customize_insert_paths(PlannerInfo *root, List *pathlist)
```

Comment:
maybe `replace_hypertable_insert_paths` is better to make it clear that we are replacing the paths.